src/QsCompiler/CompilerOptimizations/OptimizingTransformations/CallableInlining.fs

```diff
@@ -122,42 +129,87 @@ type internal CallableInliner(callables) =
 
     // The current callable we're in the process of transforming
     let mutable currentCallable: QsCallable option = None
+    let mutable renamer: VariableRenamer option = None
 
-    /// Inline an expression as a ScopeStatement, with many checks to ensure correctness.
-    /// Returns None if the expression cannot be safely inlined.
-    let safeInline expr =
+    /// The random generator to give unique variable names
+    let random = System.Random()
+
+    let tryInline expr =
         maybe {
             let! ii = tryGetInliningInfo callables expr
 
-            do! check (countReturnStatements ii.body = 0)
-            let! current = currentCallable
-            do! check (cannotReachCallable callables ii.body current.FullName)
+            let! currentCallable = currentCallable
+            let! renamer = renamer
+            renamer.clearStack()
+            do! check (cannotReachCallable callables ii.body currentCallable.FullName)
             do! check (ii.functors.controlled < 2)
             // TODO - support multiple Controlled functors
             do! check (cannotReachCallable callables ii.body ii.callable.FullName || isLiteral callables ii.arg)
 
             let newBinding = QsBinding.New ImmutableBinding (toSymbolTuple ii.specArgs, ii.arg)
-            let newStatements = ii.body.Statements.Insert (0, newBinding |> QsVariableDeclaration |> wrapStmt)
-            return {ii.body with Statements = newStatements} |> QsScopeStatement.New |> QsScopeStatement
+            let newStatements =
+                ii.body.Statements.Insert (0, newBinding |> QsVariableDeclaration |> wrapStmt)
+                |> Seq.map renamer.Scope.onStatement
+                |> Seq.map (fun s -> s.Statement)
+                |> ImmutableArray.CreateRange
+            return ii, newStatements
         }
 
+    /// Inline an expression representing a callable with no return statements.
+    /// Returns None if the expression cannot be safely inlined.
+    let safeInline expr =
+        maybe {
+            let! ii, newStatements = tryInline expr
+            do! check (countReturnStatements ii.body = 0)
```

Comment:
I'd suggest to move getting the inlining info out of tryInline, and swap the order of these two instructions to save some processing.